distributed/worker.py

```diff
@@ -1929,6 +1930,14 @@ def transition_flight_missing(
         ts.done = False
         return {}, []
 
+    def transition_fetch_missing(
+        self, ts: TaskState, *, stimulus_id: str
+    ) -> RecsInstrs:
```

Comment:
I'd like to see some assertions about `ts` in here. Likely at least
```python
assert not ts.done  # ??? really not sure about this. I find `done` confusingly named.
if self.validate:
    assert ts.state == "fetch"
    assert not ts.who_has
```

Reply:
This is in `validate_task_fetch` already, which gets called at the end of every transitions call.  I think that we're safe here.

Reply:
You mean `validate_task_missing`?

Reply:
Both validation methods assert the correct (I think) who_has state

```python
    def validate_task_fetch(self, ts):
        assert ts.key not in self.data
        assert self.address not in ts.who_has
        assert not ts.done
        assert ts in self.data_needed
        assert ts.who_has

        for w in ts.who_has:
            assert ts.key in self.has_what[w]
            assert ts in self.pending_data_per_worker[w]

    def validate_task_missing(self, ts):
        assert ts.key not in self.data
        assert not ts.who_has
        assert not ts.done
        assert not any(ts.key in has_what for has_what in self.has_what.values())
        assert ts in self._missing_dep_flight
```

So after this transition is called, the validate_task_missing method will be called, and verify that `not ts.who_has`

Reply:
Hrm, it looks like we are asserting the incoming state by default though.  

This seems a little odd to me given how few lines of code there are in between getting the state and choosing this method (below for convenience)

```python
        start = ts.state
        func = self._transitions_table.get((start, cast(str, finish)))
```

I'd be happy to add it for now as convention is you like

Reply:
Yeah, I'm also not entirely sure why convention seems to be to assert the incoming state within the transition function.

I like having some assertions for readability though. It's a little easier to read the code in the future when the expectations are laid out clearly at the top.

Reply:
It looks like it's inconsistently done.  I'm going to pass on this one unless you feel strongly about it.  If you do speak up and I'll add it.  I'm probably hesitating here just because it feels weird to keep doing something just because we've been doing it.

Reply:
> I strongly suggest not to rely on these validate calls. They are helpful but do not replace testing.

We should talk about this in more depth if you want to rely less on them.  Validation testing has, historically, been invaluable in maintaining correctness and stability.  

This is still testing.  The differnce now is that the tests themselves trigger certain behaviors, and assertions are checked in a more systematic way.  It is, I think, a better way of verifying state than explicitly checking state in every test.  This would be, I think, an inefficient way of writing our tests.

> The exception is lost in tornado and the only thing we see is an error log. Sometimes that causes the tests to get stuck but it's not reliable

If tests are passing even when these are not then that's certainly an issue and we should address it quickly.  You might not be saying this though.  If these aren't as ergonomic as we'd like then let's see if we can make them more ergonomic.

Alternatively, if we have a good alternative for the validation methods then I'm happy to engage.  I would be -1 to putting explicit state testing at this into all of the tests though.  I'm curious to learn what other alternatives there might be.  Could I ask you to raise an issue with thoughts and we can move the conversation there?

Reply:
I agree with you, they are useful and I don't want to get rid of it. I'm just saying that we cannot blindly rely on them at this point in time. The way our concurrency model with tornado works is that `AssertionError` are just lost and logged. Sometimes that causes a worker to deadlock which then is a good thing because the test times and fails. However, it depends on where this assert is exactly called and relying on this "implicit deadlock" is not great.

To counter this, I proposed https://github.com/dask/distributed/pull/4735 a while ago which proposes to log an exception and close the worker if a transition error occurs. I believe this would be a drastic behaviour but still a sane one, even for production. If anything goes wrong during state transitions, we should throw the worker away and rely on the scheduler to clean up the mess.
Thoughts?

Reply:
We only run in validation mode in tests anyway, so I'm totally fine with it.